### PR TITLE
History

### PR DESCRIPTION
**Pull request by rhigdon**

### CODE REVIEW

File: heathskin/game_state.py [hunk 1]
```diff
@@ -106,10 +116,12 @@ def get_entity_by_name(self, ent_id, default=None):
...
                 result_id = "1"
-            elif ent_id == self.friendly_user_name:
-                result_id = "2"
+            elif hasattr(self,'player1') and  ent_id == self.player1:
+                result_id = "2" if self._is_player_first() else "3"
+            elif hasattr(self, 'player2') and ent_id == self.player2:
+                result_id = "3" if self._is_player_first() else "2"
             else:
-                result_id = "3"
+                print 'failed to get entity by name : %s' % ent_id
```
Comment (amckinley) [inline heathskin/game_state.py]: raise PreventableException; make it bubble up if we got this logic wrong

File: heathskin/game_state.py [hunk 2]
```diff
@@ -96,6 +102,10 @@ def start_new_game(self):
         self.logger.info("Starting new game")
         self.entities = {}
         self.parser = LogParser(self)
+        if hasattr(self, 'player1'):
```
Comment (amckinley) [inline heathskin/game_state.py]: just declare game_state.player1 and game_state.player2 in the **init** method.

File: heathskin/game_state.py [hunk 3]
```diff
@@ -46,15 +50,17 @@ def _get_heroes_from_entities(self):
...
+        entity2 = self.entities.get('2')
+        entity3 = self.entities.get('3')
         our_hero = self.entities.get('4')
         second_pos = self.entities.get('3').get_tag('HERO_ENTITY')
         enemy_hero = self.entities.get(str(second_pos))
         if our_hero.get_tag('ZONE') == 'OPPOSING PLAY (Hero)':
             enemy_hero = our_hero
             our_hero = self.entities.get(str(second_pos))
-            self.player = self.entities.get('3')
+            self.player = entity2 if self._is_player_first() else entity3
```
Comment (amckinley) [inline heathskin/game_state.py:61]: isnt this wrong for spectated games?

File: heathskin/game_state.py [hunk 4]
```diff
@@ -38,7 +38,11 @@ def _create_history(self, *args, **kwargs):
         history.enemy_health = 30
         history.hero_health = 30
         history.turns = kwargs.get('turns')
-        history.first = self._is_player_first()
+        history.first = not self._is_player_first()
```
Comment (amckinley) [inline heathskin/game_state.py:41]: name this history.went_first?

File: heathskin/game_state.py [hunk 5]
```diff
@@ -96,6 +102,10 @@ def start_new_game(self):
         self.logger.info("Starting new game")
         self.entities = {}
         self.parser = LogParser(self)
+        if hasattr(self, 'player1'):
+            del self.player1
+        if hasattr(self, 'player2'):
+            del self.player2
```
Comment (amckinley) [inline heathskin/game_state.py]: just set them back to None

File: heathskin/game_state.py [hunk 6]
```diff
@@ -106,10 +116,12 @@ def get_entity_by_name(self, ent_id, default=None):
         except ValueError:
             if ent_id == "GameEntity":
                 result_id = "1"
-            elif ent_id == self.friendly_user_name:
-                result_id = "2"
+            elif hasattr(self,'player1') and  ent_id == self.player1:
```
Comment (amckinley) [inline heathskin/game_state.py]: no space between self,'player1', extra space between and ent_id

File: heathskin/log_parser.py [hunk 7]
```diff
@@ -209,8 +209,10 @@ def parse_debug_print_power(self, msg):
                     self.logger.debug("starting sub-action TAG_CHANGE")
                     results = self.match_tag_action(msg)
                     entity_name = results.pop('entity_name')
+                    self._setup_players_from_entity_name(entity_name)
                     target_ent = self.game_state.get_entity_by_name(entity_name)
-                    target_ent.update_tag(**results)
+                    if target_ent:
```
Comment (amckinley) [inline heathskin/log_parser.py]: remove this if after changing get_entity_by_name to raise()

File: heathskin/log_parser.py [hunk 8]
```diff
@@ -260,13 +262,25 @@ def parse_debug_print_power(self, msg):
             results = self.match_tag_action(msg)
             self.logger.debug("tag change msg: %s", results)
             entity_name = results.pop('entity_name')
+            self._setup_players_from_entity_name(entity_name)
             target_ent = self.game_state.get_entity_by_name(entity_name)
-
-            target_ent.update_tag(**results)
+            if target_ent:
```
Comment (amckinley) [inline heathskin/log_parser.py]: remove this if after changing get_entity_by_name to raise()

File: heathskin/game_state.py (hunk 1 above)
Comment (rhigdon) [inline heathskin/game_state.py]: So I am seeing plenty of lines like this 
<code>
TAG_CHANGE Entity=[name=Gul'dan id=36 zone=PLAY zonePos=0 cardId=HERO_07 player=2] tag=NUM_TURNS_IN_PLAY value=1
</code>
which would hit this exception case.. is this entity really entity 36?  does this line need further parsing?

**Comment by rhigdon** [comment]
we shouldn't merge this until we figure out this "preventable exception" issue

**Pull request merged**